Development/nmos/constraints.cpp

```diff
@@ -0,0 +1,165 @@
+            if (constraint.has_field(nmos::fields::constraint_minimum) && !subconstraint.has_field(nmos::fields::constraint_enum) && !subconstraint.has_field(nmos::fields::constraint_minimum))
+            {
+                return false;
+            }
+            // subconstraint should have maximum or enum if constraint has maximum
+            if (constraint.has_field(nmos::fields::constraint_maximum) && !subconstraint.has_field(nmos::fields::constraint_enum) && !subconstraint.has_field(nmos::fields::constraint_maximum))
+            {
+                return false;
+            }
+            if (constraint.has_field(nmos::fields::constraint_minimum) && subconstraint.has_field(nmos::fields::constraint_minimum))
+            {
+                if (constraint.at(U("minimum")).has_field(nmos::fields::numerator))
+                {
+                    if (nmos::parse_rational(constraint.at(U("minimum"))) > nmos::parse_rational(subconstraint.at(U("minimum"))))
+                    {
+                        return false;
+                    }
+                }
+                else if (nmos::fields::constraint_minimum(constraint) > nmos::fields::constraint_minimum(subconstraint))
+                {
+                    return false;
+                }
+            }
+            if (constraint.has_field(nmos::fields::constraint_maximum) && subconstraint.has_field(nmos::fields::constraint_maximum))
+            {
+                if (constraint.at(U("maximum")).has_field(nmos::fields::numerator))
+                {
+                    if (nmos::parse_rational(constraint.at(U("maximum"))) < nmos::parse_rational(subconstraint.at(U("maximum"))))
+                    {
+                        return false;
+                    }
+                }
+                else if (nmos::fields::constraint_maximum(constraint) < nmos::fields::constraint_maximum(subconstraint))
+                {
+                    return false;
+                }
+            }
+            // subconstraint enum values should match constraint
+            if (subconstraint.has_field(nmos::fields::constraint_enum))
+            {
+                const auto& subconstraint_enum_values = nmos::fields::constraint_enum(subconstraint).as_array();
+                if (subconstraint_enum_values.end() == std::find_if(subconstraint_enum_values.begin(), subconstraint_enum_values.end(), [&constraint](const web::json::value& enum_value)
+                {
+                    return match_constraint(enum_value, constraint);
+                }))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // Constraint Set B is a subset of Constraint Set A if all of Parameter Constraints of Constraint Set B, except for meta, are present in Constraint Set A and each Parameter Constraint of Constraint Set B, except for meta, is a subconstraint of the according Parameter Constraint of Constraint Set A.
```

Comment:
"Constraint Set B is a subset of Constraint Set A if all of Parameter Constraints of Constraint Set B, except for meta, are present in Constraint Set A and each Parameter Constraint of Constraint Set B, except for meta, is a subconstraint of the according Parameter Constraint of Constraint Set A."

I don't think this is right... For B to be a subset of A, it's not that all param constraints of B must be found in A, but rather that all param constraints in A must be (equally or more) constrained in B, right?

I think we can also remove the "except for meta" since meta are not parameter constraints according to BCP-004-01. "The `<category>` name `meta` is reserved for metadata related to Constraint Sets. These attributes are not constraints themselves."

Therefore:

Constraint Set B is a subset of Constraint Set A if all Parameter Constraints of Constraint Set A are present in Constraint Set B, and for each Parameter Constraint that is present in both, the Parameter Constraint of Constraint Set B is a subconstraint of the Parameter Constraint of Constraint Set A.

```suggestion
        // Constraint Set B is a subset of Constraint Set A if all Parameter Constraints of Constraint Set A are present in Constraint Set B, and for each Parameter Constraint
        // that is present in both, the Parameter Constraint of Constraint Set B is a subconstraint of the Parameter Constraint of Constraint Set A.
```

Reply:
This is a really crucial finding, thank you. I definitely must organize better integration testing to continue working on this PR.